pkg/agent/flexvolume/controller.go

```diff
@@ -108,44 +108,10 @@ func (c *FlexvolumeController) Attach(attachOpts AttachOptions, devicePath *stri
 				return fmt.Errorf("failed to create volume CRD %s. %+v", crdName, err)
 			}
 			// Some other attacher beat us in this race. Kubernetes will retry again.
-			return fmt.Errorf("failed to attach volume %s. Volume is already attached by a different pod", crdName)
+			return fmt.Errorf("failed to attach volume %s for pod %s/%s. Volume is already attached by a different pod", crdName, attachOpts.PodNamespace, attachOpts.Pod)
 		}
 	} else {
 		// Volume has already been attached.
-		// Check if there is already an attachment with RW.
-		index := getPodRWAttachmentObject(volumeattachObj)
-		if index != -1 {
-			// check if the RW attachment is orphaned.
-			attachment := &volumeattachObj.Attachments[index]
-			pod, err := c.clientset.Core().Pods(attachment.PodNamespace).Get(attachment.PodName, metav1.GetOptions{})
-			if err != nil {
-				if !errors.IsNotFound(err) {
-					return fmt.Errorf("failed to get pod CRD %s/%s. %+v", attachment.PodNamespace, attachment.PodName, err)
-				}
-
-				// Attachment is orphaned. Update attachment record and proceed with attaching
-				attachment.Node = node
-				attachment.MountDir = attachOpts.MountDir
-				attachment.PodNamespace = attachOpts.PodNamespace
-				attachment.PodName = attachOpts.Pod
-				attachment.ReadOnly = attachOpts.RW == ReadOnly
-				err = c.volumeAttachmentController.Update(volumeattachObj)
-				if err != nil {
-					return fmt.Errorf("failed to update volume CRD %s. %+v", crdName, err)
-				}
-			} else {
-				// Attachment is not orphaned. Original pod still exists. Dont attach.
-				return fmt.Errorf("failed to attach volume %s. Volume is already attached by pod %s/%s. Status %+v", crdName, attachment.PodNamespace, attachment.PodName, pod.Status.Phase)
-			}
-		} else {
-			// No RW attachment found. Check if this is a RW attachment request.
-			// We only support RW once attachment. No mixing either with RO
-			if attachOpts.RW == "rw" && len(volumeattachObj.Attachments) > 0 {
-				return fmt.Errorf("failed to attach volume %s. Volume is already attached by one or more pods", crdName)
-			}
-
-		}
-
 		// find if the attachment object has been previously created
 		found := false
 		for _, a := range volumeattachObj.Attachments {
```

Comment:
this check here for mount dir seems to be a really important part of the fencing logic for this fix that allows us to retry a failed attachment for a given pod/node/PVC.  I think the existing comment of `// find if the attachment object has been previously created` could probably be augmented a bit to explain why/how this scenario happens and why when `found = true`, this is a safe operation to proceed with attaching.

Reply:
Ok ill add a comment

Reply:
Ok ill add a comment